### ISSUE TITLE
feat(image): Research nydus/stargz

### ISSUE THREAD

**Issue by gaocegege** [tensorchord/envd#1176]
## Description

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

**Comment by xieydd** [comment]
In production enviroment, i rarely see images below 20G, some user even put the data in image. I think image accelerate is highlight and practical solutions to user problems.

**Comment by gaocegege** [comment]
Yep, I think so.

**Comment by gaocegege** [comment]
```
filename: usr/local/lib/python3.8/dist-packages/wrapt-1.14.1.dist-info/top_level.txt, offset: 1046700032, size: 6
filename: usr/local/lib/python3.8/dist-packages/zipp/, offset: 1046701056, size: 0
filename: usr/local/lib/python3.8/dist-packages/zipp/__init__.py, offset: 1046701568, size: 8659
filename: usr/local/lib/python3.8/dist-packages/zipp/__pycache__/, offset: 1046710784, size: 0
filename: usr/local/lib/python3.8/dist-packages/zipp/__pycache__/__init__.cpython-38.pyc, offset: 1046711296, size: 10762
filename: usr/local/lib/python3.8/dist-packages/zipp/__pycache__/py310compat.cpython-38.pyc, offset: 1046723072, size: 406
filename: usr/local/lib/python3.8/dist-packages/zipp/py310compat.py, offset: 1046724096, size: 309
filename: usr/local/lib/python3.8/dist-packages/zipp-3.10.0.dist-info/, offset: 1046725120, size: 0
filename: usr/local/lib/python3.8/dist-packages/zipp-3.10.0.dist-info/INSTALLER, offset: 1046725632, size: 4
filename: usr/local/lib/python3.8/dist-packages/zipp-3.10.0.dist-info/LICENSE, offset: 1046726656, size: 1050
filename: usr/local/lib/python3.8/dist-packages/zipp-3.10.0.dist-info/METADATA, offset: 1046728704, size: 3672
filename: usr/local/lib/python3.8/dist-packages/zipp-3.10.0.dist-info/RECORD, offset: 1046733312, size: 707
filename: usr/local/lib/python3.8/dist-packages/zipp-3.10.0.dist-info/WHEEL, offset: 1046734848, size: 92
filename: usr/local/lib/python3.8/dist-packages/zipp-3.10.0.dist-info/top_level.txt, offset: 1046735872, size: 5
```

The index generated by https://github.com/awslabs/soci-snapshotter

**Comment by gaocegege** [comment]
> SOCI addresses these issues by loading from the original, unmodified OCI image. Instead of converting the image, it builds a separate index artifact (the "SOCI index"), which lives in the remote registry, right next to the image itself. At container launch time, SOCI Snapshotter queries the registry for the presence of the SOCI index using the mechanism developed by the [OCI Reference Types working group](https://github.com/opencontainers/wg-reference-types).

**Comment by gaocegege** [comment]
About what the snapshot is:

https://github.com/containerd/containerd/blob/main/docs/content-flow.md

**Comment by cutecutecat** [comment]
Difference of stargz/nydus:
- https://github.com/dragonflyoss/image-service/issues/50

Design report:
- estargz: https://github.com/containerd/stargz-snapshotter/blob/main/docs/estargz.md
- nydu: https://github.com/dragonflyoss/image-service/blob/master/docs/nydus-design.md

Pros & Cons:
TODO

**Comment by gaocegege** [comment]
@cutecutecat  Are you interested in this? You can pick it up. And I'd appreciate it.

**Comment by cutecutecat** [comment]
> @cutecutecat Are you interested in this? You can pick it up. And I'd appreciate it.

Yes, I would like to pick it. Are there anything else that need to be investigated? 

As we have known buildkit can built both of them, I think I could build the large image  by `buildctl` and test the time cost and size?
- #1086 
- #51

**Comment by cutecutecat** [comment]
Nydus is conflict with `--export-cache` and `--import-cache`, is this acceptable in envd? @gaocegege 

> Since exported Nydus image will always have one more metadata layer than images in other compression types, Nydus image cannot be exported/imported as cache.

ref: https://github.com/moby/buildkit/blob/master/docs/nydus.md and https://github.com/moby/buildkit/issues/2046

**Comment by cutecutecat** [comment]
# Prefetch
> Estargz and Nydus support prefetch. This can be used to mitigate runtime performance drawbacks caused by the on-demand fetching of each file.

Maybe we could use https://github.com/docker-slim/docker-slim to do a scan of some typical ML training case, in order to pick which file is hotspot and need to be prefetched.

**Comment by gaocegege** [comment]
I think we can verify the benifits of theses tools. For example, we can run a shell in tensorflow. And see the startup time.

**Comment by cutecutecat** [comment]
`golang:1.18-alpine` is used to build and run a simple `hello.go` to test Golang building cost of stargz.
`mskwyditd/pytorch-cuda-python3.1` is used to run a simple `train.go` to test Python building cost of stargz, waiting to be finished, as I am finding a good example of model training.


Image | File | Pull | Run first time | Run second time
-- | -- | -- | -- | --
golang:1.18-alpine | hello.go | 96.4 | 1.37 | 0.4
golang:1.18-alpine + stargz | hello.go | / | 37.8 | 1.0
mskwyditd/pytorch-cuda-python3.10 | train.py | 7138.8 |   |  
mskwyditd/pytorch-cuda-python3.10 + stargz | train.py |   |   |

**Comment by gaocegege** [comment]
It's awesome!
